### PR TITLE
Qorb integration as connection pool for database

### PR DESCRIPTION
**Pull request by smklein**
Replaces all usage of bb8 with a new connection pooling library called [qorb](https://github.com/oxidecomputer/qorb).

qorb, detailed in RFD 477, provides the following benefits over bb8:
- It allows lookup of multiple backends via DNS SRV records
- It dynamically adjusts the number of connections to each bakend based on their health, and prioritizes vending out connections to healthy backends
- It should be re-usable for both our database and progenitor clients (using a different "backend connector", but the same core library and DNS resolution mechanism).

Fixes https://github.com/oxidecomputer/omicron/issues/4192
Part of https://github.com/oxidecomputer/omicron/issues/3763 (fixes CRDB portion)

### CODE REVIEW

File: nexus-config/src/postgres_config.rs
Comment (smklein) [inline nexus-config/src/postgres_config.rs:47]: I don't love that this method makes this assumption, though this appears to be embedded mostly within tests.

It's necessary within `nexus/db-queries/src/db/pool.rs` to implement `SingleHostResolver`, which is our way of identifying backends as `SocketAddr` types.

I can try to make the "address" of the backend generic through qorb, so we could leave it as a URL?

File: nexus/db-queries/src/db/collection_attach.rs
Comment (smklein) [inline nexus/db-queries/src/db/collection_attach.rs:860]: There are two flavors of `db::Pool` that are exposed now:

1. One that uses DNS to resolve backends (this will be used in prod, and recommended), and
2. One that uses a hard-coded address for resolution and skips DNS (this is used by a lot of our tests)

I'm using the function name `new_qorb_single_host` to identify case (2) explicitly.

File: nexus/src/context.rs
Comment (smklein) [inline nexus/src/context.rs:210]: We need to pull out the DNS addresses here to help bootstrap the qorb pool below.

We will eventually be able to dynamically discover DNS servers with qorb, but we need to supply a value of https://github.com/oxidecomputer/qorb/blob/bb432c9574c393bf2eb7b19516b996aa5584573e/src/resolvers/dns.rs#L387-L390 first

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [inline nexus/db-queries/src/db/pool.rs:63]: This `hardcoded_ttl` config is load-bearing. The default is `None`, and the `DnsResolver` uses a TTL from the DNS records themselves to identify the the records as present or not. **However**, all our DNS records are being saved with a TTL of "zero", so they are immediately discarded.

File: nexus-config/src/postgres_config.rs
Comment (hawkw) [inline nexus-config/src/postgres_config.rs:39]: nit, take it or leave it: I would, generally, avoid the use of "Safety" in comments that don't specifically deal with _memory_ safety. personally, i would expect to see this documentation under a heading like
```suggestion
    /// # Panics
```

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool.rs:34]: "contact the pool directly" doesn't make a ton of sense to me in this context --- I'm assuming this is for the case where we use a single DB address, such as in tests?

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool.rs:54]: not a blocker or anything, but: it occurs to me that `qorb` could provide an `impl Resolver for watch::Receiver<AllBackends>` that just clones the rx, which would kind of obviate the need for this type..

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool.rs:88]: nit, not important: personally, i'm not super convinced that these all need `_qorb_` in their names; that seems kinda like an implementation detail. as a user of this module, I feel like i mostly just care about the behavior of the pool, not which library is used to provide it.

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool.rs:137]: perhaps these should be gated behind a `test_util` feature or something, which can be enabled by dev-dependencies on this crate?

File: nexus/db-queries/src/db/pool_connection.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool_connection.rs:22]: does this need to be public? or is it internal to `db-queries`?

File: nexus/db-queries/src/db/pool_connection.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool_connection.rs:38]: nitpicky: a function taking multiple positional arguments of the same type with different semantics is always kind of a red flag to me...it seems easy to accidentally swap the ordering of the username and db. if this is going to be a public API, i might consider a struct which passes these arguments as named fields, or newtypes or something. on the other hand, if this is internal to the `db-queries` crate... :woman_shrugging:

**Review by hawkw — COMMENTED** [review]
Overall, this looks really nice! I left a few minor suggestions --- I hope they're helpful!

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool.rs:35]: Yeah, this was incorrect phrasing -- it's for cases where we want to contact the **database** directly without performing resolution first, when there's a single DB as you mentioned.

Comment (smklein) [reply nexus/db-queries/src/db/pool.rs:35]: (I'll update the comment)

File: nexus-config/src/postgres_config.rs
Comment (smklein) [reply nexus-config/src/postgres_config.rs:39]: Sure thing, will update

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool.rs:55]: Hrmmm I suppose so, but the choice of resolver still needs to be caller-controlled. I think we could probably adjust the API of `Pool::new` to take a `watch::Receiver<AllBackends>` instead of a Boxed `Resolver`, but they feel pretty similar functionally, right?

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool.rs:91]: Fair enough, I think it was way more relevant during implementation when I had a combination of bb8 and qorb (briefly) but that is no longer the case. I'll update these names.

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool.rs]: Good idea -- this is a little harder to do on `new_single_host`, because of the current clients and nexus configs, but I can definitely do this for the failfast one at least

File: nexus/db-queries/src/db/pool_connection.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool_connection.rs:23]: I can make it crate-local! Done

File: nexus/db-queries/src/db/pool_connection.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool_connection.rs]: Updated to use a struct explicitly (even though it is, as you suggested, crate-internal)

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [reply nexus/db-queries/src/db/pool.rs:93]: :woman_shrugging: it's up to you, i don't feel terribly strongly about this!

Comment (smklein) [reply nexus/db-queries/src/db/pool.rs:93]: I think it was the right call! Updated 'em

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool.rs:128]: nit, take it or leave it: maybe instead of "quickly" we should just say the actual duration?

File: nexus/db-queries/src/db/pool.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool.rs:130]: it might be nice if this comment explained when test code should use `new_single_host` and when it should use this constructor?

**Review by hawkw — APPROVED** [review]
Overall, this looks good to me! Up to you whether it's worth waiting for a review from someone else, as well.

File: nexus/db-queries/src/db/pool_connection.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool_connection.rs]: - https://crates.io/crates/uri looks deprecated, so probably not that?
- https://docs.rs/url/2.5.2/url/ looks popular , and like it could be used for this purpose, though it doesn't have exactly that API you described
- https://docs.rs/http/latest/http/uri/struct.Builder.html also looks like it's a builder for URLs

I'm gonna go with the `url` crate, and use it primarily for the query parameters.

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool.rs]: Updated (it's one millisecond)

File: nexus/db-queries/src/db/pool.rs
Comment (smklein) [reply nexus/db-queries/src/db/pool.rs]: Sure thing, added.

File: nexus/db-queries/src/db/pool_connection.rs
Comment (hawkw) [inline nexus/db-queries/src/db/pool_connection.rs:68]: i really thought there was a builder-style interface for constructing URLs from parts like this, but looking at the `url` docs, i think it must be the _other_ Rust URL library that has that...

**Comment by smklein** [comment]
This is still pretty much ready to go, but I'm interested in getting https://github.com/oxidecomputer/omicron/pull/5912 resolved first so we don't fragment our DNS client usage -- especially considering that we've been bumping into mysterious timeouts with hickory DNS that still need resolution.

**Comment by davepacheco** [comment]
#5912 has now been resolved and the hickory-dns timeouts understood so I think this can move forward.

**Pull request merged**
